src/OpenTelemetry.Contrib.Instrumentation.GrpcCore/RpcScope.cs

```diff
@@ -0,0 +1,231 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RpcScope{TRequest, TResponse}" /> class.
+        /// </summary>
+        /// <param name="fullServiceName">Full name of the service.</param>
+        /// <param name="recordMessageEvents">if set to <c>true</c> [record message events].</param>
+        protected RpcScope(string fullServiceName, bool recordMessageEvents)
+        {
+            this.FullServiceName = fullServiceName?.TrimStart('/') ?? "unknownservice/unknownmethod";
+            this.recordMessageEvents = recordMessageEvents;
+        }
+
+        /// <summary>
+        /// Gets the full name of the service.
+        /// </summary>
+        protected string FullServiceName { get; }
+
+        /// <summary>
+        /// Records a request message.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        public void RecordRequest(TRequest request)
+        {
+            this.requestMessageCounter++;
+
+            if (this.activity == null || !this.activity.IsAllDataRequested || !this.recordMessageEvents)
+            {
+                return;
+            }
+
+            this.AddMessageEvent(typeof(TRequest).Name, request as IMessage, request: true);
+        }
+
+        /// <summary>
+        /// Records a response message.
+        /// </summary>
+        /// <param name="response">The response.</param>
+        public void RecordResponse(TResponse response)
+        {
+            this.responseMessageCounter++;
+
+            if (this.activity == null || !this.activity.IsAllDataRequested || !this.recordMessageEvents)
+            {
+                return;
+            }
+
+            this.AddMessageEvent(typeof(TResponse).Name, response as IMessage, request: false);
+        }
+
+        /// <summary>
+        /// Completes the RPC.
+        /// </summary>
+        public void Complete()
+        {
+            if (this.activity == null)
+            {
+                return;
+            }
+
+            // The overall Span status should remain unset however the grpc status code attribute is required
+            this.StopActivity((int)Grpc.Core.StatusCode.OK);
+        }
+
+        /// <summary>
+        /// Records a failed RPC.
+        /// </summary>
+        /// <param name="exception">The exception.</param>
+        public void CompleteWithException(Exception exception)
+        {
+            if (this.activity == null)
+            {
+                return;
+            }
+
+            var grpcStatusCode = Grpc.Core.StatusCode.Unknown;
+            var description = exception.Message;
+
+            if (exception is RpcException rpcException)
+            {
+                grpcStatusCode = rpcException.StatusCode;
+                description = rpcException.Message;
+            }
+
+            this.StopActivity((int)grpcStatusCode, description);
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()

```

Comment:
nit: This pattern of having Complete + Dispose is a bit strange, IMO. Why not let Dispose always stop the activity and then have the other methods just modify the state/tags?

Reply:
I'm going to push back on this one. I would argue that the methods exposed by RpcScope (RecordMessage, Complete + CompleteWithException) fit more naturally with the various reading/writing proxies used. Additionally, gRPC Core itself doesn't really utilize Dispose like that. From the docs of one of the AsyncServerStreamingCall for e.g.

```
        //
        // Summary:
        //     Provides means to cleanup after the call. If the call has already finished normally
        //     (response stream has been fully read), doesn't do anything. Otherwise, requests
        //     cancellation of the call which should terminate all pending async operations
        //     associated with the call. As a result, all resources being used by the call should
        //     be released eventually.
        //
        // Remarks:
        //     Normally, there is no need for you to dispose the call unless you want to utilize
        //     the "Cancel" semantics of invoking Dispose.
```

Reply:
It's internal, so not a big deal. What I was trying to say (more or less) was: [Dispose has a specific purpose and a well-defined pattern](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose) so seeing it co-opted for some other purpose is a code smell.